### ISSUE TITLE
Bug (I think): Wagtail userbar displays in language of page content, not Wagtail admin language

### ISSUE THREAD

**Issue by Scotchester** [wagtail/wagtail#3991]
### Issue Summary

I was working on [one of our Spanish-language pages](https://www.consumerfinance.gov/es/prepararse-para-comprar-casa/) yesterday, and I noted that when I opened up the userbar that the text there was displayed in Spanish:

<img width="248" alt="screen shot 2017-11-02 at 09 34 32" src="https://user-images.githubusercontent.com/1044670/32379716-b2476586-c084-11e7-919f-c53cbe339073.png">

I was expecting that the userbar would be in English, as the main Wagtail admin is in English, even when editing pages whose content is in other languages. While I am not an expert in Django internationalization, I think I understand why this is happening, but it would be nice if there were a way to override it so that it always matches the language of the main Wagtail admin.

### Steps to Reproduce

1. In your English-language Wagtail admin, create a page and change the language on its Settings tab to Spanish
1. View the live page and open the userbar
1. Observe that it is in Spanish

### Technical details

* Python version: 2.7.10
* Django version: 1.8.15
* Wagtail version: 1.10.1 (yes, this is a current source of shame for us 😞)

**Comment by chosak** [comment]
I agree that this behavior seems somewhat unintuitive - I'd expect the Wagtail language to be consistent throughout regardless of the language of the page being viewed.

Could this be accomplished by wrapping the userbar template in a Django [`language` template tag](https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#switching-language-in-templates) using `get_current_language`?

**Comment by gasman** [comment]
We need to pick apart the division of responsibilities a little here... I'm assuming you're using the wagtailtrans package, because Wagtail itself has no concept of a page having a defined language (and so "changing the language on its Settings tab to Spanish" is not something that exists on a vanilla Wagtail install).

I suspect there's at least one way this bug manifests itself without wagtailtrans: if a user sets a language preference in the Wagtail admin (footer menu -> account settings -> language preferences), it won't be respected by the userbar, because the language selection is applied [using a decorator on `/admin` views](https://github.com/wagtail/wagtail/blob/6dc89c1f7f7d846594d093f1e52efcf9d38ab20c/wagtail/wagtailadmin/decorators.py#L28-L29). This specific manifestation of the bug is probably easy enough to fix - we just need to move that logic somewhere where the `{% wagtailuserbar %}` tag can re-use it.

However, digging a bit deeper, it looks like we've not been particularly rigorous about what happens when the user's language preference is unset. [On one code path](https://github.com/wagtail/wagtail/blob/6dc89c1f7f7d846594d093f1e52efcf9d38ab20c/wagtail/wagtailusers/models.py#L42-L43) - where a UserProfile exists but the language field is blank - we default to the global `settings.LANGUAGE_CODE`; if there's no UserProfile at all, we fall through and let [Django's own language-selection logic](https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#how-django-discovers-language-preference) handle it. This will _often_ be the same as falling back on `settings.LANGUAGE_CODE`, but there are a few special cases that could throw things off: one of them is Django's `LocaleMiddleware`, which might select another language based on session/cookie/Accept-Language header data. Another one is the middleware from wagtailtrans :-(

The "probably good enough" fix is for the userbar to check `UserProfile.preferred_language` and fall back on `settings.LANGUAGE_CODE`, bypassing any middleware that would normally have a bearing on the chosen language. The proper fix - which might not even be possible - would need to dig in to the internals of those middlewares, and try to distinguish between the ones like `LocaleMiddleware`, which we _do_ want the admin to respect, and the ones like wagtailtrans which are only meant to apply to the front-end page.

**Comment by laymonage** [comment]
An update on this as I thought #13047 was related.

As of the last comment, the fallback have been changed from `settings.LANGUAGE_CODE` to respect  `LocaleMiddleware` by using `get_language()` in #4845.

I think the "good enough fix" described above and implemented in https://github.com/wagtail/wagtail/pull/6220/commits/1ad240e19afa26c4112c94d1b1ff05775d43eed7?w=1#r474667293 was enough. And, as explained, I'm not sure there's a way to fix it without digging too much into `LocaleMiddleware`'s internals.

If you never intend for the content's language to be used in the admin, the userbar now respects the `WAGTAILADMIN_PERMITTED_LANGUAGES` setting as of #13047. In this issue's example, this means you can use that setting to not enable the Spanish language in the admin. If you do wish to enable Spanish in the admin but have the userbar not affected by the content's locale, then users must set their preferred language in the admin, which I think is fair game.
